pkg/internal/api/mutation/registry_test.go

```diff
@@ -228,7 +228,7 @@ func TestMutate(t *testing.T) {
 				},
 			},
 			expErr:   false,
-			expPatch: []byte(`[{"op":"replace","path":"/spec/request","value":"bXV0YXRpb24gY2FsbGVk"}]`),
+			expPatch: []byte(`[{"op":"add","path":"/spec/request","value":"bXV0YXRpb24gY2FsbGVk"}]`),
```

Comment:
Thanks for highlighting the relevant part of the spec.
Having read that specification of the Add operation, it makes sense that jsonpatch should use it here, since the original object `testCRBytes` generated by https://github.com/jetstack/cert-manager/blob/7e51eec2b0ba5211e5280ff7a6eacef996676bd1/test/unit/gen/certificaterequest.go#L28-L36 would have an empty `spec.request`....although....

Then I went looking at https://github.com/jetstack/cert-manager/blob/7e51eec2b0ba5211e5280ff7a6eacef996676bd1/pkg/apis/certmanager/v1/types_certificaterequest.go#L95-L97 where I see that the Request field is not marked `// +optional` and `omitempty`, so I think it would end up with a `null` value when serialized to JSON.

Which means that replace might be the appropriate operation here.

Is that what https://github.com/gomodules/jsonpatch/pull/29 is about?

And do you agree that the `CR.Spec.Request` field should be marked `// +optional` and `omitempty`?

Reply:
> And do you agree that the CR.Spec.Request field should be marked // +optional and omitempty?

I am not sure about that.

Regarding `replace` vs `add`, replace will work only if the original key exists. `add` is more forgiving. `add` will add the new key or replace a key if already exists. So, both `add` and `replace` will work correctly in this case, since the key exists but the value is `null`. 

`gomodules/jsonpatch` will generate `add` if the original key was missing or `null`. It will generate `replace` if the original key exists and not `null`. Since this library is correct within the json patch spec, I would like to leave it as-is. Also, this current version is also used in the controller-runtime. So, I rather leave it as-is. That's why I updated the tests to match the new expected behavior.
`

Reply:
You'll think I'm being difficult, but the point I'm getting hung up on is that the original will serialize to:

```json

{
  "kind": "CertificateRequest",
  "apiVersion": "cert-manager.io/v1",
  "metadata": {
    "name": "test-cr",
    "namespace": "default-unit-test-ns",
    "creationTimestamp": null
  },
  "spec": {
    "issuerRef": {
      "name": ""
    },
    "request": null
  },
  "status": {}
}

```
...in which `request` key *does* exist (with a null value), so `replace` seems like the correct jsonpatch operation, according to the spec.
By "correct", I mean the most specific operation. 
The advantage as I see it is that  if the original has been changed and no longer has the key, when the patch is being applied, then the `Replace` operation will fail and the patch will be rejected....at which point the client can get the latest version of the target object and recalculate the patch.

You argue that `Add` will work here too, and I agree, but my question then is, what is the point of `Replace` operation? Why is it included in the jsonpatch spec? When should it be used if not here?

When the change in  https://github.com/gomodules/jsonpatch/pull/29  is released and used in cert-manager, then this test will once again fail won't it? 

Seems to me that the `Request` field should be `omitempty` with a `// +kubebuilder:validation:Required` marker, but that doesn't seem to work properly either: https://github.com/jetstack/cert-manager/pull/4028

Reply:
> You argue that Add will work here too, and I agree, but my question then is, what is the point of Replace operation? Why is it included in the jsonpatch spec? When should it be used if not here?

`replace` can change the data type of a key from array to scalar types. `add` seems to just add/replace (for scalar types) or append (for array).